### PR TITLE
Resurrect processing's raster calculator

### PR DESCRIPTION
**Pull request by nirvn**
## Description
@nyalldawson , @alexbruy , this ~~ressurects~~ resurrects the processing raster calculator, and improves the algorithm logic in several ways.

The big logic improvement is around "reference layers". While in the past, the algorithm would consider _all_ raster layers listed in the expression widget, the updated logic allows for users to select which layer(s) to use as reference to create the output extent and cellsize. This offers a more fine-grained control over the output.

The algorithm also allows for no reference layers to be provided if all of the needed output values are provided (extent, cellsize, and CRS). The user can also leave some values to be determined by the reference layer(s), and specify others manually.

Also, the algorithm now supports multiple CRSs while calculating output extent, cellsize, and CRS. That's _really_ nice, and crucial to make all of the extent parameter's function (such as draw extent on canvas) work properly.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/plugins/processing/algs/qgis/RasterCalculator.py [hunk 1]
```diff
@@ -118,10 +119,44 @@ def displayName(self):
     def processAlgorithm(self, parameters, context, feedback):
         expression = self.parameterAsString(parameters, self.EXPRESSION, context)
         layers = self.parameterAsLayerList(parameters, self.LAYERS, context)
-        layersDict = {}
+
         if layers:
+            layersDict = {}
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/RasterCalculator.py]: this change looks wrong ?

Comment (nirvn) [reply python/plugins/processing/algs/qgis/RasterCalculator.py]: As in, this layersDict = {} definitions should be moved out of the conditional statement?

Comment (nyalldawson) [reply python/plugins/processing/algs/qgis/RasterCalculator.py]: yes - it's referenced elsewhere so should always be made

Comment (nirvn) [reply python/plugins/processing/algs/qgis/RasterCalculator.py]: Fixed

**Comment by alexbruy** [comment]
Looks good to me

**Pull request merged**